devdoc/waiter_queue_requirements.md

```diff
@@ -0,0 +1,103 @@
+# `waiter_queue` requirements
+
+## Overview
+
+`waiter_queue` is a module that allows consumers to park calls and for producers to signal that the parked calls be executed. A "parked call" consists of a callback function and a context. The consumer provides the callback function and context when it parks a call. The consumer expects the parked call's callback function to be executed with the provided context when the producer signals.
+
+`waiter_queue` has two parts: `waiter_queue_ll` that contains the logic for parking and unblocking calls and `waiter_queue_hl` that contains logic for making `waiter_queue` thread-safe.
```

Comment:
Yes, the user can call `waiter_queue_hl_add_waiter` from an `UNBLOCK_CALLBACK`.

Reply:
So will the real implementation not use srw_lock? Because srw_lock is not reentrant

Reply:
Hmm that's a good point.

Reply:
`add_waiter` during an  ongoing `unblock_waiters` shouldn't be a problem, because only parked calls that were parked before the ongoing `unblock_waiters` will be executed.

`unblock_waiters` during an ongoing `unblock_waiters` can create problems because now we have don't have any guarantees that the parked calls will be executed with the data given to the first `unblock_waiters` call before being executed with the data from the second `unblock_waiters` call. But the user cannot expect that anyway if the two `unblock_waiters` calls are being made concurrently.

What are your thoughts?